velox/exec/PartitionedOutput.h

```diff
@@ -72,6 +75,19 @@ class Destination {
   void
   serialize(const RowVectorPtr& input, vector_size_t begin, vector_size_t end);
 
+  // Sets the next target size for flushing. This is called at the
+  // start of each batch of output for the destination. The effect is
+  // to make different destinations ready at slightly different times
+  // so that for an even distribution of output we avoid a bursty
+  // traffic pattern where all consumers contend for the network at
+  // the same time. This is done for each batch so that the average
+  // batch size for each converges.
+  void setTargetSizePct() {
+    // Flush at  70 to 120% of target row or byte count.
+    targetSizePct_ = 70 + (folly::Random::rand32(rng_) % 50);
+    targetNumRows_ = (10000 * targetSizePct_) / 100;
```

Comment:
Is the new `targetNumRows_` added here to cap each output to about `10K` rows?   
Because this is hard-coded, is the intention here to set a hard limit, or just try its best to cap?    

I noticed that for a single-partition scenario (there will only be one `IndexRange{0, numRows}` in `rows_`), this restriction will not take effect.

Reply:
@lingbin Interesting question. There are 2 limits: bytes and number of rows. 

```
  // Limit serialized pages to 1MB.
  static const uint64_t kMaxPageSize = 1 << 20;
  const uint64_t maxPageSize = std::max<uint64_t>(
      kMinDestinationSize,
      std::min<uint64_t>(kMaxPageSize, maxBufferedBytes_ / numDestinations_));

    if (bytesInCurrent_ >= adjustedMaxBytes ||
        row_ - firstRow >= targetNumRows_) {
```

> I noticed that for a single-partition scenario (there will only be one IndexRange{0, numRows} in rows_), this restriction will not take effect.

I don't see this being the case. single-partition means one Destination, but not necessarily one row. Would you clarify a bit?

Reply:
@mbasmanova   In the current code implementation, `targetNumRows_` actually limits the number of `IndexRange` in `rows_`(has been renamed to `ranges_`).

https://github.com/facebookincubator/velox/blob/36f9621f3ffb65c4600438b612ade3aaaa0e2ff2/velox/exec/PartitionedOutput.cpp#L49-L50

- In non-single-partition scenarios, each `IndexRange` contains only one row, that is, the number of `IndexRange` in `rows_` is equal to the actual number of rows, so there is no problem.

- But for the single-partition scenario (using the `Destination::addRows()` interface instead of `Destination::addRow()`), the result is that there will **only be one `IndexRange` element in `rows_`** (where includes all rows), so it cannot be restricted by `targetNumRows_`.

So I'm wondering why we need to limit to ~10K rows? And is it intended to be a hard limit, or just try its best? And if this constraint is broken in a single-partition scenario, will there be any problems?

Reply:
@lingbin 

You are correct. There is a TODO in Destination::advance:

```
    // TODO: add support for serializing partial ranges if the full range is too
    // big.
```

> I'm wondering why we need to limit to ~10K rows? 

This limit appears arbitrary. 

PartitionedOutputBufferManager::getData() allows the caller (downstream task/worker) to limit the total size of the response. This is only possible if PartitionedOutput limits the size of the serialized pages it adds via PartitionedOutputBufferManager::enqueue(). PartitionedOutput::getOutput() has a hard-coded limit of 1MB, which allows POBM to respect the maxSize at the granularity of 1MB. The limit on number of rows is not needed.

```
  // Limit serialized pages to 1MB.
  static const uint64_t kMaxPageSize = 1 << 20;
  const uint64_t maxPageSize = std::max<uint64_t>(
      kMinDestinationSize,
      std::min<uint64_t>(kMaxPageSize, maxBufferedBytes_ / numDestinations_));
```